### PR TITLE
Fallback to opam 2.1 when encountering solver timeouts

### PR DESCRIPTION
**Pull request by kit-ty-kate**
Temporary alternative to https://github.com/ocurrent/opam-repo-ci/pull/65
Proposal by @dra27

### CODE REVIEW

File: lib/opam_build.ml [hunk 1]
```diff
@@ -21,6 +21,11 @@ let opam_install ~variant ~upgrade_opam ~pin ~with_tests ~pkg =
         opam remove -y %s && opam %s%s %s
         res=$?
         test "$res" = 0 && exit 0
+        if test "$res" = 60 && diff -q /usr/bin/opam /usr/bin/opam-2.0; then
+          sudo cp /usr/bin/opam-2.1 /usr/bin/opam
```
Comment (avsm) [inline lib/opam_build.ml:25]: I believe this should be a hard link and not a cp (as that's what the base dockerfiles do)

**Comment by dra27** [comment]
Just to check I've understood the script correctly: if opam 2.0 fails with a solver timeout, but opam 2.1 succeeds, the CI step will still fail with `exit 1`, yes? (I'm hoping that's true: i.e. the build still fails, but we can extract two pieces of info rather than one)

File: lib/opam_build.ml [hunk 2]
```diff
@@ -21,6 +21,11 @@ let opam_install ~variant ~upgrade_opam ~pin ~with_tests ~pkg =
         opam remove -y %s && opam %s%s %s
         res=$?
         test "$res" = 0 && exit 0
+        if test "$res" = 60 && diff -q /usr/bin/opam /usr/bin/opam-2.0; then
```
Comment (dra27) [inline lib/opam_build.ml:24]:
```suggestion
        if test "$res" != 0 && diff -q /usr/bin/opam /usr/bin/opam-2.0; then
```
?? Can this also be hit with depext-related errors too?

File: lib/opam_build.ml (hunk 1 above)
Comment (kit-ty-kate) [reply lib/opam_build.ml]: Fixed.

File: lib/opam_build.ml (hunk 2 above)
Comment (kit-ty-kate) [reply lib/opam_build.ml:24]: Testing seems to indicates that `opam-depext` also returns exit code 60 properly: https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/597b4467d6799043fc07b4eafee2d7a038f95821/variant/compilers,4.04,mirage-conduit.3.0.0

**Comment by kit-ty-kate** [comment]
> Just to check I've understood the script correctly: if opam 2.0 fails with a solver timeout, but opam 2.1 succeeds, the CI step will still fail with `exit 1`, yes? (I'm hoping that's true: i.e. the build still fails, but we can extract two pieces of info rather than one)

No it will ignore the error and just go with whatever opam 2.1 ends up doing.
The rational is that the solver timeout failures won't be fixed in opam 2.0 and above all we need testing the underlying package, not "knowing that this package makes the solver fail in CI". Solver timeouts can't be fixed (or very very very rarely) in the packages themselves so there is nothing we can do about it.
Continuing displaying these errors are mostly useless because they are so common that as opam maintainers we end up just ignoring them and there are so many of them on big runs that is makes detecting the real genuine failures more of a pain than it should.

File: lib/opam_build.ml (hunk 2 above)
Comment (dra27) [reply lib/opam_build.ml:24]: That wasn't quite what I meant - is solver timeout the only issue which can occur? I thought there were depext-related problems?

**Comment by dra27** [comment]
- opam 2.0 has a solver timeout, so `$res` at L24 is 60 and you trigger your new branch
- opam 2.1 succeeds so `$res` will be `0` at L27
- L30 then runs `test "$res" != 31 && exit 1` and so fails the step

Did I miss something?

**Comment by dra27** [comment]
If the fallback is to run opam 2.1 and ignore the fact that 2.0 failed, then this largely becomes a switch to opam 2.1 "by the backdoor".

> The rational is that the solver timeout failures won't be fixed in opam 2.0

Knowing that opam 2.0 is failing a lot of PRs increases the chance of the fixes (CUDF trimming, etc.) being back-ported!

I wasn't proposing that you start getting green ticks for solver timeouts, rather that if there's another error further down the line which 2.1 can reveal that we can have that instead.

File: lib/opam_build.ml (hunk 2 above)
Comment (kit-ty-kate) [reply lib/opam_build.ml:24]: ah, you're correct. Though, I would say I'd prefer to take one problem at a time. The depext issue is more difficult to detect and isolate.

**Comment by kit-ty-kate** [comment]
>     * opam 2.0 has a solver timeout, so `$res` at L24 is 60 and you trigger your new branch
> 
>     * opam 2.1 succeeds so `$res` will be `0` at L27
> 
>     * L30 then runs `test "$res" != 31 && exit 1` and so fails the step
> 
> 
> Did I miss something?



>     * opam 2.0 has a solver timeout, so `$res` at L24 is 60 and you trigger your new branch
> 
>     * opam 2.1 succeeds so `$res` will be `0` at L27
> 
>     * L30 then runs `test "$res" != 31 && exit 1` and so fails the step
> 
> 
> Did I miss something?

oops, sorry about that, my bad.

**Comment by dra27** [comment]
Excuse the lag, I went head down on OCaml's CI yesterday!

Sorry to push back on this, but the behaviour of the script before was what I was suggesting - if opam 2.0 is timing out, then it cannot be the case that CI passes the PR. The problem at the moment is that the solver timeout means you don't know if there's an actual problem with the package/revdep (i.e. a build problem) because you actually get to the build.

In other words, for example, looking at https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/769d51a59bbb306dea0cccd468f8480c5167bea8 (PR#18107): you presently have:

- 4.04
  x mirage-conduit.3.0.0 (failed: Sorry, resolution of the request timed out.)
    + tests (not started)

I'm suggest that opam 2.1 can be used to allow, for example, the tests to be started and run, but there must at the end be _at least_ the solver timeout error. It might want enhancing to say "Sorry, **opam 2.0** timed out resolving the request" and it may also be the the "Summary of errors" section wants to have a separate group for solver timeouts, so that it's easier for maintainers to see the important errors, but the error _cannot_ disappear because users may well see it.

**Comment by kit-ty-kate** [comment]
My original rational was:
- timeouts have been there at least for the past 3 years and back then the answer I got on this is "it's a hard problem to fix in opam"
- opam 2.1 worked and ultimately it's the thing that is going to end-up being used by users
- I haven't seen any complaints from users outside CI about timeouts

I didn't really think the fix could be backported to 2.0, however Louis seem to agree with you so I opened https://github.com/ocaml/opam/pull/4538

--------

Other than that I agree on your idea, however the annoying thing is that tests of will end up not being tested as we skip failures and it is difficult to make a distinction in the code between "failure" and "failure but pass to the next". Maybe I'm ignoring something in the code that would help me do that though, I'll have a look.

**Comment by kit-ty-kate** [comment]
I think it looks good enough now. I'll merge it tomorrow.

**Pull request merged**